File: api/common/event_stream_manager.py
import json
import os
from dataclasses import asdict
from enum import Enum
from typing import Any, Dict, List, Union

import redis.asyncio as redis

from .user_message import ChatInteraction
from .status_update import StatusUpdate


class EventType(str, Enum):
    CHAT_INTERACTION = "chat_interaction"
    STATUS_UPDATE = "status_update"


class EventStreamManager:
    def __init__(self, redis_host: str = None, redis_port: int = None):
        self.redis_host = redis_host or os.getenv("REDIS_HOST", "localhost")
        self.redis_port = redis_port or int(os.getenv("REDIS_PORT", "6379"))
        self.redis_client = redis.Redis(
            host=self.redis_host,
            port=self.redis_port,
            decode_responses=True,
        )

    def _stream_key(self, workflow_id: str) -> str:
        return f"events:{workflow_id}"

    async def append_chat_interaction(self, workflow_id: str, chat_interaction: ChatInteraction) -> int:
        return await self._append(workflow_id, EventType.CHAT_INTERACTION, chat_interaction)

    async def append_status_update(self, workflow_id: str, status_update: StatusUpdate) -> int:
        return await self._append(workflow_id, EventType.STATUS_UPDATE, status_update)

    async def _append(self, workflow_id: str, event_type: EventType, obj: Union[ChatInteraction, StatusUpdate]) -> int:
        event = {"type": event_type.value, "content": asdict(obj)}
        return await self.redis_client.rpush(self._stream_key(workflow_id), json.dumps(event))

    async def get_events_from_index(self, workflow_id: str, from_index: int = 0) -> List[Dict[str, Any]]:
        raw = await self.redis_client.lrange(self._stream_key(workflow_id), from_index, -1)
        events = []
        for item in raw:
            try:
                events.append(json.loads(item))
            except json.JSONDecodeError:
                continue
        return events

    async def delete_stream(self, workflow_id: str) -> bool:
        deleted = await self.redis_client.delete(self._stream_key(workflow_id))
        return deleted > 0

    async def close(self):
        await self.redis_client.aclose()

File: src/temporal_supervisor/redis/event-stream-manager.ts
import Redis from 'ioredis';
import { ChatInteraction, StatusUpdate } from '../shared';

const REDIS_HOST = process.env.REDIS_HOST ?? 'localhost';
const REDIS_PORT = parseInt(process.env.REDIS_PORT ?? '6379', 10);

export type EventType = 'chat_interaction' | 'status_update';

export interface ChatInteractionEvent {
  type: 'chat_interaction';
  content: ChatInteraction;
}

export interface StatusUpdateEvent {
  type: 'status_update';
  content: StatusUpdate;
}

export type StreamEvent = ChatInteractionEvent | StatusUpdateEvent;

/**
 * TypeScript port of Python's EventStreamManager.
 * Same Redis key format and JSON event structure so the Python API can read it.
 *
 * Key: events:{workflowId}   (Redis list, RPUSH / LRANGE)
 */
export class EventStreamManager {
  private client: Redis;

  constructor(host = REDIS_HOST, port = REDIS_PORT) {
    this.client = new Redis({ host, port, lazyConnect: true });
  }

  private streamKey(workflowId: string): string {
    return `events:${workflowId}`;
  }

  async appendChatInteraction(workflowId: string, interaction: ChatInteraction): Promise<number> {
    const event: ChatInteractionEvent = { type: 'chat_interaction', content: interaction };
    return this.client.rpush(this.streamKey(workflowId), JSON.stringify(event));
  }

  async appendStatusUpdate(workflowId: string, update: StatusUpdate): Promise<number> {
    const event: StatusUpdateEvent = { type: 'status_update', content: update };
    return this.client.rpush(this.streamKey(workflowId), JSON.stringify(event));
  }

  async getEventsFromIndex(workflowId: string, fromIndex = 0): Promise<StreamEvent[]> {
    const raw = await this.client.lrange(this.streamKey(workflowId), fromIndex, -1);
    const events: StreamEvent[] = [];
    for (const item of raw) {
      try {
        events.push(JSON.parse(item) as StreamEvent);
      } catch {
        // skip malformed
      }
    }
    return events;
  }

  async deleteStream(workflowId: string): Promise<boolean> {
    const deleted = await this.client.del(this.streamKey(workflowId));
    return deleted > 0;
  }

  async close(): Promise<void> {
    await this.client.quit();
  }
}
